packages/ui/src/component/interactive/UIInteractive.ts

```diff
@@ -0,0 +1,276 @@
+    for (let i = 0, n = transitions.length; i < n; i++) {
+      const transition = transitions[i];
+      if (transition instanceof type) {
+        results.push(transition);
+      }
+    }
+    return results;
+  }
+
+  /**
+   * Get transition which match the type.
+   * @param type - The type of the transition
+   * @returns	The first transition which match type
+   */
+  getTransition<T extends Transition>(type: new (interactive: UIInteractive) => T): T | null {
+    const transitions = this._transitions;
+    for (let i = 0, n = transitions.length; i < n; i++) {
+      const transition = transitions[i];
+      if (transition instanceof type) {
+        return transition;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Add transition based on the transition type.
+   * @param type - The type of the transition
+   * @returns	The transition which has been added
+   */
+  addTransition<T extends new (interactive: UIInteractive) => Transition>(type: T): InstanceType<T> {
+    const transition = new type(this) as InstanceType<T>;
+    this._transitions.push(transition);
+    transition._setState(this._state, true);
+    return transition;
+  }
+
+  override onUpdate(deltaTime: number): void {
+    this.globalInteractive && this._transitions.forEach((transition) => transition._onUpdate(deltaTime));
+  }
+
+  override onPointerBeginDrag(): void {
+    this._isPointerDragging = true;
+    this._updateState(false);
+  }
+
+  override onPointerEndDrag(): void {
+    this._isPointerDragging = false;
+    this._updateState(false);
+  }
+
+  override onPointerEnter(): void {
+    this._isPointerInside = true;
+    this._updateState(false);
+  }
+
+  override onPointerExit(): void {
+    this._isPointerInside = false;
+    this._updateState(false);
+  }
+
+  override onDestroy(): void {
+    super.onDestroy();
+    this._transitions.forEach((transition) => transition.destroy());
+  }
+
+  /**
+   * @internal
+   */
+  _getRootCanvas(): UICanvas {
+    this._isRootCanvasDirty && Utils.setRootCanvas(this, Utils.searchRootCanvasInParents(this));
+    return this._rootCanvas;
+  }
+
+  /**
+   * @internal
+   */
+  _getGroup(): UIGroup {
+    this._isGroupDirty && Utils.setGroup(this, Utils.searchGroupInParents(this));
+    return this._group;
+  }
+
+  // @ts-ignore
+  override _onEnableInScene(): void {
+    // @ts-ignore
+    super._onEnableInScene();
+    Utils.setRootCanvasDirty(this);
+    Utils.setGroupDirty(this);
+    this._updateState(true);
+  }
+
+  // @ts-ignore
+  override _onDisableInScene(): void {
+    // @ts-ignore
+    super._onDisableInScene();
+    Utils.cleanRootCanvas(this);
+    Utils.cleanGroup(this);
+    this._isPointerInside = this._isPointerDragging = false;
+  }
+
+  /**
+   * @internal
+   */
+  _removeTransition(transition: Transition): void {
+    const transitions = this._transitions;
+    for (let i = transitions.length - 1; i >= 0; i--) {
+      if (transitions[i] === transition) {
+        transitions.splice(i, 1);
```

Comment:
Replace i that you want remove to last, and then delete last transition may be better performance